File: src/main/java/ru/org/linux/topic/TopicListService.java
/*
 * Copyright 1998-2022 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.topic;

import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import ru.org.linux.group.Group;
import ru.org.linux.section.Section;
import ru.org.linux.tag.TagNotFoundException;
import ru.org.linux.tag.TagService;
import ru.org.linux.user.User;
import ru.org.linux.user.UserErrorException;

import javax.annotation.Nullable;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

@Service
public class TopicListService {
  private static final Logger logger = LoggerFactory.getLogger(TopicListService.class);

  private final TagService tagService;

  private final TopicListDao topicListDao;

  public TopicListService(TagService tagService, TopicListDao topicListDao) {
    this.tagService = tagService;
    this.topicListDao = topicListDao;
  }

  /**
   * Получение списка топиков.
   *
   * @param section секция
   * @param group   группа
   * @param tag     тег
   * @param offset  смещение в результатах выборки
   * @param year    год
   * @param month   месяц
   * @return список топиков
   * @throws UserErrorException
   * @throws TagNotFoundException
   */
  public List<Topic> getTopicsFeed(
          Section section,
          Group group,
          String tag,
          Integer offset,
          Integer year,
          Integer month,
          int count,
          @Nullable User currentUser
  ) throws TagNotFoundException {
    logger.debug(
            "TopicListService.getTopicsFeed()" +
                    "; section=" + ((section != null) ? section.toString() : "(null)") +
                    "; group=" + ((group != null) ? group.toString() : "(null)") +
                    "; tag=" + tag +
                    "; offset=" + offset +
                    "; year=" + year +
                    "; month=" + month
    );

    TopicListDto topicListDto = new TopicListDto();

    if (section != null) {
      topicListDto.setSection(section.getId());
      if (section.isPremoderated()) {
        topicListDto.setCommitMode(TopicListDao.CommitMode.COMMITED_ONLY);
      } else {
        topicListDto.setCommitMode(TopicListDao.CommitMode.POSTMODERATED_ONLY);
      }
    }

    if (group != null) {
      topicListDto.setGroup(group.getId());
    }

    if (tag != null) {
      topicListDto.setTag(tagService.getTagId(tag, false));
    }

    if (month != null && year != null) {
      topicListDto.setDateLimitType(TopicListDto.DateLimitType.BETWEEN);
      Calendar calendar = Calendar.getInstance();

      calendar.set(year, month - 1, 1, 0, 0, 0);
      topicListDto.setFromDate(calendar.getTime());

      calendar.add(Calendar.MONTH, 1);
      topicListDto.setToDate(calendar.getTime());
    } else {
      topicListDto.setLimit(count);
      topicListDto.setOffset(offset > 0 ? offset : null);
      if (tag == null && group == null && !section.isPremoderated()) {
        topicListDto.setDateLimitType(TopicListDto.DateLimitType.FROM_DATE);
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(new Date());
        calendar.add(Calendar.MONTH, -6);
        topicListDto.setFromDate(calendar.getTime());
      }
    }
    return topicListDao.getTopics(topicListDto, currentUser);
  }

  /**
   * Получение списка топиков пользователя.
   *
   * @param user       объект пользователя
   * @param offset     смещение в результатах выборки
   * @param isFavorite true если нужно выбрать избранные сообщения пользователя
   * @return список топиков пользователя
   */
  public List<Topic> getUserTopicsFeed(User user, Integer offset, boolean isFavorite, boolean watches) {
    return getUserTopicsFeed(user, null, null, offset, isFavorite, watches);
  }

  /**
   * Получение списка топиков пользователя.
   *
   * @param user      объект пользователя
   * @param section   секция, из которой выбрать сообщения
   * @param offset    смещение в результатах выборки
   * @param favorites true если нужно выбрать избранные сообщения пользователя
   * @return список топиков пользователя
   */
  public List<Topic> getUserTopicsFeed(User user, Section section, Group group, Integer offset, boolean favorites, boolean watches) {
    logger.debug(
            "TopicListService.getTopicsFeed()" +
                    "; user=" + ((user != null) ? user.toString() : "(null)") +
                    "; section=" + section +
                    "; offset=" + offset +
                    "; isFavorite=" + favorites
    );

    TopicListDto topicListDto = new TopicListDto();
    topicListDto.setLimit(20);
    topicListDto.setOffset(offset);
    topicListDto.setCommitMode(TopicListDao.CommitMode.ALL);
    topicListDto.setUserId(user.getId());
    topicListDto.setUserFavs(favorites);
    topicListDto.setUserWatches(watches);
    if (section != null) {
      topicListDto.setSection(section.getId());
    }
    if (group != null) {
      topicListDto.setGroup(group.getId());
    }

    return topicListDao.getTopics(topicListDto, null);
  }

  /**
   * Получение списка черновиков пользователя.
   *
   * @param user   объект пользователя
   * @param offset смещение в результатах выборки
   * @return список топиков пользователя
   */
  public List<Topic> getDrafts(User user, Integer offset) {
    TopicListDto topicListDto = new TopicListDto();
    topicListDto.setLimit(20);
    topicListDto.setOffset(offset);
    topicListDto.setCommitMode(TopicListDao.CommitMode.ALL);
    topicListDto.setUserId(user.getId());
    topicListDto.setShowDraft(true);

    return topicListDao.getTopics(topicListDto, null);
  }


  /**
   * Получение списка топиков для RSS-ленты.
   *
   * @param section  секция
   * @param group    группа
   * @param fromDate от какой даты получить список
   * @param noTalks  без Talks
   * @param tech     только технические
   * @return список топиков для RSS-ленты
   */
  public List<Topic> getRssTopicsFeed(
          Section section,
          Group group,
          Date fromDate,
          boolean noTalks,
          boolean tech
  ) {
    logger.debug(
            "TopicListService.getRssTopicsFeed()" +
                    "; section=" + ((section != null) ? section.toString() : "(null)") +
                    "; group=" + ((group != null) ? group.toString() : "(null)") +
                    "; fromDate=" + fromDate +
                    "; noTalks=" + noTalks +
                    "; tech=" + tech
    );

    TopicListDto topicListDto = new TopicListDto();

    if (section != null) {
      topicListDto.setSection(section.getId());
    }
    if (group != null) {
      topicListDto.setGroup(group.getId());
    }

    topicListDto.setDateLimitType(TopicListDto.DateLimitType.FROM_DATE);
    topicListDto.setFromDate(fromDate);

    topicListDto.setNotalks(noTalks);
    topicListDto.setTech(tech);
    topicListDto.setLimit(100);

    if (section.isPremoderated()) {
      topicListDto.setCommitMode(TopicListDao.CommitMode.COMMITED_ONLY);
    } else {
      topicListDto.setCommitMode(TopicListDao.CommitMode.POSTMODERATED_ONLY);
    }
    return topicListDao.getTopics(topicListDto, null);
  }

  public List<Topic> getUncommitedTopic(Section section, Date fromDate, boolean includeAnonymous) {
    TopicListDto topicListDto = new TopicListDto();
    topicListDto.setCommitMode(TopicListDao.CommitMode.UNCOMMITED_ONLY);
    if (section != null) {
      topicListDto.setSection(section.getId());
    }

    topicListDto.setDateLimitType(TopicListDto.DateLimitType.FROM_DATE);
    topicListDto.setFromDate(fromDate);
    topicListDto.setIncludeAnonymous(includeAnonymous);

    return topicListDao.getTopics(topicListDto, null);
  }

  public List<DeletedTopic> getDeletedTopics(int sectionId, boolean skipEmptyReason, boolean includeAnonymous) {
    return topicListDao.getDeletedTopics(sectionId, skipEmptyReason, includeAnonymous);
  }

  public List<Topic> getMainPageFeed(boolean showGalleryOnMain, int count, boolean hideMinor) {
    TopicListDto topicListDto = new TopicListDto();

    topicListDto.setLimit(count);

    topicListDto.setDateLimitType(TopicListDto.DateLimitType.FROM_DATE);
    // лучше бы этот лимит был по commit_date на главной
    topicListDto.setFromDate(DateTime.now().minusMonths(3).toDate());

    if (hideMinor) {
      topicListDto.setMiniNewsMode(TopicListDto.MiniNewsMode.MAJOR);
    }

    topicListDto.setCommitMode(TopicListDao.CommitMode.COMMITED_ONLY);

    if (showGalleryOnMain) {
      topicListDto.setSection(Section.SECTION_NEWS, Section.SECTION_GALLERY);
    } else {
      topicListDto.setSection(Section.SECTION_NEWS);
    }

    return topicListDao.getTopics(topicListDto, null);
  }


  /**
   * Корректировка смещения в результатах выборки.
   *
   * @param offset оригинальное смещение
   * @return скорректированное смещение
   */
  public int fixOffset(Integer offset) {
    if (offset != null) {
      if (offset < 0) {
        return 0;
      }

      if (offset > 200) {
        return 200;
      }

      return offset;
    } else {
      return 0;
    }
  }

  public List<Topic> getTopics(TopicListDto topicListDto, @Nullable User currentUser) {
    return topicListDao.getTopics(topicListDto, currentUser);
  }

  public List<DeletedTopic> getDeletedUserTopics(User user, int topics) {
    return topicListDao.getDeletedUserTopics(user, topics);
  }
}

File: src/main/scala/ru/org/linux/topic/TopicListService.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.topic

import org.springframework.stereotype.Service
import ru.org.linux.auth.{AnySession, NonAuthorizedSession}
import ru.org.linux.group.Group
import ru.org.linux.section.{Section, SectionService}
import ru.org.linux.tag.TagService
import ru.org.linux.topic.TopicListRequest.CommitMode
import ru.org.linux.topic.TopicListRequest.DateLimit
import ru.org.linux.user.User

import java.util.Calendar
import java.util.Date
import java.time.{Instant, ZonedDateTime}

@Service
object TopicListService:
  val MaxOffset = 300

  def fixOffset(offset: Int): Int =
    if offset < 0 then
      0
    else if offset > MaxOffset then
      MaxOffset
    else
      offset

@Service
class TopicListService(tagService: TagService, topicListDao: TopicListDao, sectionService: SectionService):
  def getTopicsFeed(
      section: Section,
      group: Option[Group],
      tag: Option[String],
      offset: Int,
      yearMonth: Option[(Int, Int)],
      count: Int,
      noTalks: Boolean,
      tech: Boolean)(using currentUser: AnySession): collection.Seq[Topic] =
    val commitMode =
      if section.isPremoderated then
        CommitMode.CommittedOnly
      else
        CommitMode.PostmoderatedOnly

    val base = TopicListRequest(
      notalks = noTalks,
      tech = tech,
      sections = Set(section.id),
      commitMode = commitMode,
      group = group.map(_.id).getOrElse(0),
      tag = tag.map(tagService.getTagId(_)).getOrElse(0)
    )

    val dto =
      if yearMonth.isDefined then
        val calendar = Calendar.getInstance
        calendar.set(yearMonth.get._1, yearMonth.get._2 - 1, 1, 0, 0, 0)
        val fromDate = calendar.getTime
        calendar.add(Calendar.MONTH, 1)
        val toDate = calendar.getTime
        base.copy(dateLimit = DateLimit.Between(fromDate, toDate))
      else
        val withLimit = base.copy(limit = Some(count), offset = Some(TopicListService.fixOffset(offset)).filter(_ > 0))
        if tag.isEmpty && group.isEmpty && !section.isPremoderated then
          val calendar = Calendar.getInstance
          calendar.setTime(new Date)
          calendar.add(Calendar.MONTH, -6)
          withLimit.copy(dateLimit = DateLimit.FromDate(calendar.getTime))
        else
          withLimit

    topicListDao.getTopics(dto, currentUser)

  def getUserTopicsFeed(
      user: User,
      section: Option[Section] = None,
      offset: Int,
      favorites: Boolean,
      watches: Boolean): collection.Seq[Topic] =
    val dto = TopicListRequest(
      limit = Some(20),
      offset = Some(offset),
      commitMode = CommitMode.All,
      userId = user.id,
      userFavs = favorites,
      userWatches = watches,
      sections = section.map(s => Set(s.id)).getOrElse(Set.empty)
    )

    topicListDao.getTopics(dto, NonAuthorizedSession)

  def getDrafts(user: User, offset: Int): collection.Seq[Topic] =
    val dto = TopicListRequest(
      limit = Some(20),
      offset = Some(offset),
      commitMode = CommitMode.All,
      userId = user.id,
      showDraft = true)

    topicListDao.getTopics(dto, NonAuthorizedSession)

  def getRssTopicsFeed(
      section: Section,
      group: Option[Group],
      fromDate: Instant,
      noTalks: Boolean,
      tech: Boolean): collection.Seq[Topic] =
    val commitMode =
      if section.isPremoderated then
        CommitMode.CommittedOnly
      else
        CommitMode.PostmoderatedOnly

    val dto = TopicListRequest(
      sections = Set(section.id),
      group = group.map(_.id).getOrElse(0),
      dateLimit = DateLimit.FromDate(Date.from(fromDate)),
      notalks = noTalks,
      tech = tech,
      limit = Some(30),
      commitMode = commitMode
    )

    topicListDao.getTopics(dto, NonAuthorizedSession)

  def getUncommitedTopic(section: Option[Section], fromDate: Date): collection.Seq[Topic] =
    val dto = TopicListRequest(
      commitMode = CommitMode.UncommittedOnly,
      sections = section.map(s => Set(s.id)).getOrElse(Set.empty),
      dateLimit = DateLimit.FromDate(fromDate))

    topicListDao.getTopics(dto, NonAuthorizedSession)

  def getDeletedTopics(sectionId: Int, skipBadReason: Boolean): Seq[DeletedTopic] =
    topicListDao.getDeletedTopics(sectionId, skipBadReason = skipBadReason)

  def getMainPageFeed(count: Int)(using session: AnySession): collection.Seq[Topic] =
    val sections =
      if session.profile.showGalleryOnMain then
        Set(Section.News, Section.Gallery, Section.Polls, Section.Articles)
      else
        Set(Section.News)

    val dto = TopicListRequest(
      limit = Some(count),
      dateLimit = DateLimit.FromDate(Date.from(ZonedDateTime.now.minusMonths(3).toInstant)),
      commitMode = CommitMode.CommittedOnly,
      sections = sections
    )

    topicListDao.getTopics(dto, NonAuthorizedSession)

  def getTopics(topicListRequest: TopicListRequest)(using currentUser: AnySession): collection.Seq[Topic] =
    topicListDao.getTopics(topicListRequest, currentUser)

  def getDeletedUserTopics(user: User, topics: Int): Seq[DeletedTopic] = topicListDao.getDeletedUserTopics(user, topics)

  def getUserSections(user: User): Seq[Section] = topicListDao.getUserSections(user).map(sectionService.idToSection)
